### PR TITLE
Aj7010

### PR DESCRIPTION
**Pull request by NicolasJourden**
I have a Philips sensor probe: the Aj7010. I used it to test the decoding of the data in different situations.
In this request you can find the code to decode the data produced by that probe.
Let me know your remarks so that the merge can be accepted.

### CODE REVIEW

**Comment by zuckschwerdt** [comment]
Can you rebase this to current master (then force push to this PR)?

**Comment by zuckschwerdt** [comment]
Are you sure the payload is a raw ADC value? Never seen that on commercial sensors. Please upload a table (txt.zip) with codes and readings (like the 5 you already have in the comment, but more, preferably 50-100).

**Comment by zuckschwerdt** [comment]
Ah, well your regression is effectively: val / 258 -20. I'd say that's 256.
But it would likely be half or quarter degrees only, with the remaining bits as checksum.

**Comment by NicolasJourden** [comment]
Yes. I have merged with master.

**Comment by NicolasJourden** [comment]
The trick with that sensor is that its display is refreshed more often than the data it transmits. We can see the byte transmitted over the air not changing while the displayed value on the sensor screen evolves.
I did not manage to get the CRC right yet. I have a large data set I can share.

**Comment by zuckschwerdt** [comment]
Just upload a zipped txt with the data here or send it to me by mail, I'll have a look.

File: include/rtl_433.h [hunk 1]
```diff
@@ -8,12 +8,32 @@
...
+#include <getopt.h>
+#endif
 
 #define DEFAULT_SAMPLE_RATE     250000
 #define DEFAULT_FREQUENCY       433920000
 #define DEFAULT_HOP_TIME        (60*10)
-#define DEFAULT_ASYNC_BUF_NUMBER    0 // Force use of default value (librtlsdr default: 15)
-#define DEFAULT_BUF_LENGTH      (16 * 32 * 512) // librtlsdr default
+#define DEFAULT_HOP_EVENTS      2
+#define DEFAULT_ASYNC_BUF_NUMBER    0 // Force use of default value (was : 32)
```
Comment (Mindavi) [inline include/rtl_433.h]: Why is this file changed? Why is the comment changed that said `// Force use of default value (librtlsdr default: 15)` to `// Force use of default value (was : 32)`? It's not obvious to me why that's changed and why the 'previous' value is now 32 instead of 15.

File: include/rtl_433.h [hunk 2]
```diff
@@ -8,12 +8,32 @@
...
+#define DEFAULT_BUF_LENGTH      (16 * 16384)
```
Comment (Mindavi) [inline include/rtl_433.h]: This gives the same total as the other `DEFAULT_BUF_LENGTH`.

File: include/rtl_433_devices.h [hunk 3]
```diff
@@ -1,7 +1,3 @@
-/** @file
-    Declaration of all available decoders.
```
Comment (Mindavi) [inline include/rtl_433_devices.h]: I don't think this should be removed.

File: src/devices/philips_AJ3650.c [hunk 4]
```diff
@@ -136,7 +136,7 @@ static int philips_callback(r_device *decoder, bitbuffer_t *bitbuffer)
     battery_status = packet[PHILIPS_PACKETLEN - 1] & 0x40;
 
     data = data_make(
-                     "model",         "",            DATA_STRING, _X("Philips-Temperature","Philips outdoor temperature sensor"),
+                     "model",         "",            DATA_STRING, _X("Philips-Temperature","Philips AJ3650 outdoor temperature sensor"),
```
Comment (Mindavi) [inline src/devices/philips_AJ3650.c]: Maybe this needs to be changed to something like Philips-AJ3650-Temperature. This is also done for other devices, e.g. https://github.com/merbanan/rtl_433/blob/6b6b89321b0cca261819a6694dec21f30995b82c/src/devices/digitech_xc0324.c#L122
@zuckschwerdt should this be changed?

If this is left as-is, the decoder for the aj3650 and aj7010 willl both report the same `model`.

File: src/devices/philips_AJ7010.c [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+ * c - channel: 0x5A=channel 1, 0x45=channel 2, 0x36=channel 3 (7 bits)
+ * t - temperature in ADC value that is then converted to deg. C. (14 bits)
+ * s - CRC: non-standard CRC-?, poly 0x?, init 0x?
+ *  
+ * Copyright (C) 2018 Nicolas Jourden <nicolas.jourden@laposte.net>
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
```
Comment (Mindavi) [inline src/devices/philips_AJ7010.c]: The rtl_433 project is licensed under the GPL-2.0 only license. Are those licenses compatible?

File: src/devices/philips_AJ7010.c [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+    if (bitbuffer->bits_per_row[0] != PHILIPS_BITLEN) {
+        if (decoder->verbose > 1) {
+            fprintf(stderr, "%s: wrong number of bits (%d)\n",
+                    __func__, bitbuffer->bits_per_row[0]);
+        }
+        return 0;
+    }
+
+    // Copy data of the packet to work on:
+    memcpy(packet, bitbuffer->bb[0]+1, PHILIPS_DATALEN);
```
Comment (Mindavi) [inline src/devices/philips_AJ7010.c]: The data is never changed, right? Don't think there's a need to make a copy, just make a pointer. E.g. `uint8_t *packet = bitbuffer->bb[0]+1` (does the +1 result in bitbuffer->bb[1]?).

File: src/devices/philips_AJ7010.c [hunk 7]
```diff
@@ -0,0 +1,164 @@
...
+    // Invert the data:
+    bitbuffer_invert(bitbuffer); // ~ on data.
+
+    // Correct number of rows?
+    if (bitbuffer->num_rows != 1) {
+        if (decoder->verbose > 1) {
+            fprintf(stderr, "%s: wrong number of rows (%d)\n",
+                    __func__, bitbuffer->num_rows);
+        }
+        return 0;
```
Comment (Mindavi) [inline src/devices/philips_AJ7010.c]: For the return codes you could look at this:
https://github.com/merbanan/rtl_433/blob/6b6b89321b0cca261819a6694dec21f30995b82c/include/r_device.h#L24-L33

I guess this would result in DECODE_ABORT_LENGTH or DECODE_ABORT_EARLY

File: include/rtl_433.h (hunk 1 above)
Comment (Mindavi) [reply include/rtl_433.h]: Oh, does this have something to do with the merge? If so, disregard the comments I've made. It just seemed strange any changes were done to this file.

**Comment by zuckschwerdt** [comment]
I had a quick look with Pulseview now. It seems the first three pulses are syncs with half the short width. The packet gap is 39ms. That way this flex decoder should work:

    rtl_433 -K FILE -R 0 -X 'n=Aj7010,m=OOK_PWM,s=2000,l=6000,y=1000,g=9900,r=40000' *.cu8

**Comment by zuckschwerdt** [comment]
@NicolasJourden I have rebased this PR as @Mindavi noted the odd changes from merge conflicts.
You probably need to `git reset --hard master ; git pull` to get these updates.

**Comment by zuckschwerdt** [comment]
I can update and merge this. But we still need samples and a table of raw codes to figure out the CRC. @NicolasJourden can you upload samples and raw codes (bitbuffer output) here?

**Comment by zuckschwerdt** [comment]
Oh sorry, I already have the samples. I'll merge and continue work in a few days.

**Comment by NicolasJourden** [comment]
Sorry we just settle home with our newborn ... giving less time to play :)

Let me know if you need anything.
I see I have made odd changes using previous version of the code, sorry for that.

**Comment by zuckschwerdt** [comment]
I have merged this with some updates as dbe6e58. But it's default disabled and needs more work.

**Comment by klohner** [comment]
I think I made some progress on the CRC in this signal.

Decoding samples in `rtl_433_tests/tests/Philips-AJ7010/01` with: 
`rtl_433 -K FILE -R 0 -F csv -X n=AJ7010,m=OOK_PWM,s=2000,l=6000,y=1000,g=10000,r=40000,invert -r *.cu8`

```
time,msg,codes,tag,model,count,num_rows,rows
@0.475136s,,{40}005a972ae7;{40}005a972acd;{40}005a972ae7,g001_433.92M_250k.cu8,AJ7010,3,3,;;
@0.475140s,,{40}005a592a29;{40}005a592a03;{40}005a592a29,g014_433.92M_250k.cu8,AJ7010,3,3,;;
@0.475148s,,{40}005a872af7;{40}005a872add;{40}005a872af7,g005_433.92M_250k.cu8,AJ7010,3,3,;;
@0.475152s,,{40}005a752a05;{40}005a752a2f;{40}005a752a05,g010_433.92M_250k.cu8,AJ7010,3,3,;;
@0.475180s,,{40}005a052679;{40}005a05265f;{40}005a052679,boot_ch1_17c_433.92M_250k.cu8,AJ7010,3,3,;;
@0.475184s,,{40}00451e267d;{40}00451e265b;{40}00451e267d,boot_ch2_17c_433.92M_250k.cu8,AJ7010,3,3,;;
@0.475188s,,{40}005a7c290f;{40}005a7c2926;{40}005a7c290f,21.4_ch1_C_433.92M_250k.cu8,AJ7010,3,3,;;
@0.479128s,,{40}005a122a62;{40}005a122a48;{40}005a122a62,g016_433.92M_250k.cu8,AJ7010,3,3,;;
@0.479172s,,{40}005a7c2a0c;{40}005a7c2a26;{40}005a7c2a0c,g006_433.92M_250k.cu8,AJ7010,3,3,;;
@0.479176s,,{40}005a8c2afc;{40}005a8c2ad6;{40}005a8c2afc,g004_433.92M_250k.cu8,AJ7010,3,3,;;
@0.479180s,,{40}005a642a14;{40}005a642a3e;{40}005a642a14,g012_433.92M_250k.cu8,AJ7010,3,3,;;
@0.479184s,,{40}005a5e2a2e;{40}005a5e2a04;{40}005a5e2a2e,g013_433.92M_250k.cu8,AJ7010,3,3,;;
@0.479204s,,{40}0036e325f0;{40}0036e325d5;{40}0036e325f0,boot_ch3_17c_433.92M_250k.cu8,AJ7010,3,3,;;
@0.479232s,,{40}005a772904;{40}005a77292d;{40}005a772904,21.4_ch1_B_433.92M_250k.cu8,AJ7010,3,3,;;

```

For each signal, it seems the first and third rows are the same, and the second differs by just the last byte.

Looking at the first row in each signal and feeding to [reveng](http://reveng.sourceforge.net/), it is able to solve:

```
$ reveng -w 8 -l -F -s 005a972ae7 005a592a29 005a872af7 005a752a05 005a052679 00451e267d 005a7c290f 005a122a62 005a7c2a0c 005a8c2afc 005a642a14 005a5e2a2e 0036e325f0 005a772904
width=8  poly=0x01  init=0x00  refin=true  refout=true  xorout=0x00  check=0x31  residue=0x00  name=(none)

```
The second row in each signal is a bit strange, though.  It seems like the last byte (the "CRC" byte) should be first XOR'd with the previous byte in the row to get a the same "true" CRC found in the first and third row.  So, for example, `21.4_ch1_B_433.92M_250k.cu8` returns rows `{40}005a772904;{40}005a77292d;{40}005a772904`.  Start with the second row, take the last byte `2d` (the "CRC" code), XOR with `29` (the previous byte in the row) to get `04` which is the true CRC code for the data (the same CRC value as found in the first and third rows).

**Comment by zuckschwerdt** [comment]
Good find. I didn't think the last byte was data, but lack of randomness and those 2nd rows threw me off. Nice to confirm that it's a check. Poly 1 is a "degenerate" CRC that's just parity.

**Comment by zuckschwerdt** [comment]
MIC added and default enabled with b863e8b. Now we just need to fine tune the temperature conversion and maybe figure out what's up with the "channel" bits.
